### PR TITLE
Add plugin for import.meta proposal

### PR DESCRIPTION
**Pull request by jkrems**
Fixes https://github.com/babel/babylon/issues/539

| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | yes
| Deprecations?     | no
| Spec compliancy?  | yes
| Tests added/pass? | yes
| Fixed tickets     | #539
| License           | MIT

Relatively straight-forward: Parses `import.meta` as a new `MetaProperty`. The expression is valid in modules but not in scripts. It can be enabled even when `dynamicImport` isn't enabled.

### CODE REVIEW

**Comment by jkrems** [comment]
The CI failure *seems* to be unrelated to this change..?

**Comment by bakkot** [comment]
Should make sure to disallow `import.meta = val`. I don't think it's specced yet, but it will be.

**Comment by jkrems** [comment]
Added tests to verify that:

1. `import.meta` is not a valid assignment target
2. `import.meta.prop` is a valid assignment target

File: test/fixtures/experimental/import-meta/without-dynamic-import/actual.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+const x = import.meta;
+const url = import.meta.url;
+import.meta;
+import.meta.url;
```
Comment (hzoo) [inline test/fixtures/experimental/import-meta/without-dynamic-import/actual.js]: Was there a reason this test didn't also have ` import.meta.couldBeMutable = true;`?

Comment (jkrems) [inline test/fixtures/experimental/import-meta/without-dynamic-import/actual.js]: No deeper reason than "order of copy&paste". ^^ Copied the `actual.js`/`expected.json` over again to keep them in sync.

File: src/parser/expression.js [hunk 2]
```diff
@@ -594,6 +598,16 @@ export default class ExpressionParser extends LValParser {
...
   }
 
+  parseImportMetaProperty(): N.MetaProperty {
+    const node = this.startNode();
+    const id = this.parseIdentifier(true);
+    this.expect(tt.dot);
+    if (this.options.sourceType !== "module") {
+      this.raise(id.start, "import.meta can only be used in modules");
+    }
+    return this.parseMetaProperty(node, id, "meta");
```
Comment (Kovensky) [inline src/parser/expression.js]: The error message if the meta property isn't `.meta` is confusing (it's hardcoded to mention `new`)

Comment (jkrems) [inline src/parser/expression.js]: Oops, good point. Added a test that checks for that & made it use the correct name.

File: src/parser/expression.js [hunk 3]
```diff
@@ -588,12 +592,22 @@ export default class ExpressionParser extends LValParser {
...
     }
 
     return this.finishNode(node, "MetaProperty");
   }
 
+  parseImportMetaProperty(): N.MetaProperty {
+    const node = this.startNode();
+    const id = this.parseIdentifier(true);
+    this.expect(tt.dot);
+    if (this.options.sourceType !== "module") {
```
Comment (existentialism) [inline src/parser/expression.js]: Can also do `if (!this.inModule)` here

File: src/parser/expression.js [hunk 4]
```diff
@@ -588,12 +592,22 @@ export default class ExpressionParser extends LValParser {
...
+      this.raise(id.start, "import.meta can only be used in modules");
```
Comment (existentialism) [inline src/parser/expression.js]: May want to make this consistent with other sourceType errors?

> 'import.meta' can only be used with 'sourceType: module'

File: src/parser/expression.js (hunk 3 above)
Comment (jkrems) [inline src/parser/expression.js]: Weird, I thought I tried that first but `.inModule` was only true top-level. Must have missed something. You're right, definitely works fine.

File: src/parser/expression.js (hunk 4 above)
Comment (jkrems) [inline src/parser/expression.js]: Heh, good point. Copied the wording from the "can only be used in functions". Should've taken the wording from import/export statements. Fixed.

**Comment by rwaldron** [comment]
This might be irrelevant, or possibly a subject to have in a new issue, but I'm curious about the drops in coverage? For example, it says above that six files were impacted, but four of them aren't changed _at all_: 

- src/parser/comments.js	87.87% <0%> (-6.07%)	⬇️
- src/tokenizer/state.js	97.14% <0%> (-2.86%)	⬇️
- src/plugins/flow.js	97.35% <0%> (-1.04%)	⬇️
- src/tokenizer/index.js	97.41% <0%> (-1%)	⬇️

**Comment by jkrems** [comment]
@rwaldron Yeah, I was wondering about the same thing. I tried rebasing on master in case it was using some weird merge head but afaict that wasn't it.

**Pull request merged**